students/chelsea_nayan/lesson09/cli_main.py

```diff
@@ -0,0 +1,167 @@
+    dataset = [max_donor_names, max_total_given, max_num_gifts, max_average_gift]
+    return dataset
+
+# Print the header
+def header_report(dataset):
+    heading = ("Donor Name".ljust(dataset[0], ' ') + "|  Total Given ".ljust(dataset[1]+3, ' ') + "|  Num Gifts ".ljust(dataset[2], ' ')+ "|  Average Gift".ljust(dataset[3]+3, ' '))
+    separation = ("-"*len(heading))
+    header = heading +  "\n" + separation
+    return header
+
+# Sort the donor list by the sum of their donations, from highest total to least
+# Make an iterable of the donor name and then a list of their donations
+def sort_report(donors):
+    sorted_dict = sorted(donors.items(), key=lambda elem: sum(elem[1]), reverse=True)
+    return sorted_dict
+
+def report():
+    # Sort donor list by sum of their donations (high to low)
+    sort_report(donors)
+
+    # Create separate lists of the four donation summary columns
+    donor_names, total_given, num_gifts, average_gift = [], [], [], []
+    for sublist in sort_report(donors):
+        donor_names.append(sublist[0])
+        total_given.append(f"{sum(sublist[1]):.2f}") # Format to two decimals
+        num_gifts.append(len(sublist[1]))
+        average_gift.append(f"{sum(sublist[1])/len(sublist[1]):.2f}") # Format to two decimals
+
+    dataset = setup_report(donor_names, total_given, num_gifts, average_gift)
+
+    # Print the header
+    print(header_report(dataset))
+
+    # Print the summary tables
+    for i in range(len(donor_names)):
+        print(f"{donor_names[i].ljust(dataset[0],' ')}  $ {str(total_given[i]).rjust(dataset[1], ' ')} {str(num_gifts[i]).rjust(dataset[2], ' ')}  $ {str(average_gift[i]).rjust(dataset[3], ' ')} ")
+
+# ------------ Send letters to everyone ---------------- #
+
+def text_send(name, amount):
+    text = f'''
+    Dear {name},
+
+        Thank you for your super, duper total donation of ${float(amount):.2f}.
+        I will buy so many things for myself.
+
+            You're the best,
+                - Chelsea
+    '''
+    return text
+
+# Generate a thank you letter and write it to a disk as a text file
+def send():
+    print('\nSending letters...')
+
+    for name in donors.keys():
+        amount = sum(donors.get(name))
+        txt_name = name.replace(" ", "_")
+        output = (f'{txt_name}.txt')
+        with open(output, 'w') as f:
+            f.write(text_send(name, amount))
+
+    print('\nSent the letters!')
+
+# ------------ Quit ---------------- #
+
+# Exit the program
+def quit():
+    print("See ya later, gater!")
+    sys.exit(0)
+
+
+# ------------ Run the Main Program ---------------- #
+
+if __name__=='__main__':
+
+    # Switch
+    arg_dict = {1: thankyou,
+                2: report,
+                3: send,
+                4: quit
+    }
+
+    # A loop to prompt the user for a response of one of four options
+    while True:
+        response = input('''
+    Choose one of four actions:
+    [1] Send a Thank You
```

Comment:
Getting an error when trying to add a donor - throws the TypeError from Donor() class in mailroom_oo.py